devicemodel/hw/pci/virtio/virtio_gpio.c

```diff
@@ -0,0 +1,942 @@
+	}
+
+	memset(&info, 0, sizeof(info));
+	rc = ioctl(fd, GPIO_GET_CHIPINFO_IOCTL, &info);
+	if (rc < 0) {
+		DPRINTF("Can't ioctl gpio device: %s, error: %s\n",
+				path, strerror(errno));
+		goto fail;
+	}
+
+	chip->lines = calloc(1, info.lines * sizeof(*chip->lines));
+	if (!chip->lines) {
+		DPRINTF("Alloc chip lines error, %s:%d, error %s\n",
+				path, chip->ngpio, strerror(errno));
+		goto fail;
+	}
+	chip->fd = fd;
+	chip->ngpio = info.lines;
+	strncpy(chip->name, info.name, sizeof(chip->name) - 1);
+	strncpy(chip->label, info.label, sizeof(chip->label) - 1);
+	strncpy(chip->dev_name, name, sizeof(chip->dev_name) - 1);
+
+	/* initialize all lines of the chip */
+	for (i = 0; i < chip->ngpio; i++) {
+		line = &chip->lines[i];
+		line->offset = i;
+		line->chip = chip;
+
+		/*
+		 * The line's fd and voffset will be initialized
+		 * when virtual gpio line connects to the real line.
+		 */
+		line->fd = -1;
+		line->voffset = -1;
+
+		/* Set line state and name via ioctl*/
+		native_gpio_update_line_info(line);
+	}
+
+	return 0;
+
+fail:
+	if (fd > 0)
+		close(fd);
+	chip->fd = -1;
+	chip->ngpio = 0;
+	return -1;
+}
+
+static int
+native_gpio_get_offset(struct native_gpio_chip *chip, char *name)
+{
+	int rc;
+	int i;
+
+	/* try to find a gpio index by offset or name */
+	if (isalpha(name[0])) {
+		for (i = 0; i < chip->ngpio; i++) {
+			if (!strcmp(chip->lines[i].name, name))
+				return i;
+		}
+	} else if (isdigit(name[0])) {
+		rc = dm_strtoi(name, NULL, 10, &i);
+		if (rc == 0 && i < chip->ngpio)
+			return i;
+	}
+	return -1;
+}
+
+static struct gpio_line *
+native_gpio_find_line(struct native_gpio_chip *chip, const char *name)
+{
+	int offset;
+	char *b, *o, *c;
+	struct gpio_line *line = NULL;
+
+	b = o = strdup(name);
+	c = strsep(&o, "=");
+
+	/* find the line's offset in the chip by name or number */
+	offset = native_gpio_get_offset(chip, c);
+	if (offset < 0) {
+		DPRINTF("the %s line has not been found in %s chip\n",
+				c, chip->dev_name);
+		goto out;
+	}
+
+	line = &chip->lines[offset];
+	if (line->busy || native_gpio_open_line(line, 0, 0) < 0) {
+		line = NULL;
+		goto out;
+	}
+
+	/* If the user sets the name of the GPIO, copy it to vname */
+	if (o)
+		strncpy(line->vname, o, sizeof(line->vname) - 1);
+
+out:
+	free(b);
+	return line;
+}
+
+
+static int
+native_gpio_init(struct virtio_gpio *gpio, char *opts)
+{
+	struct gpio_line *line;
+	char *cstr, *lstr, *tmp, *b, *o;
+	int rc;
+	int cn = 0;
+	int ln = 0;
+
+	/*
+	 * -s <slot>,virtio-gpio,<gpio resources>
+	 * <gpio resources> format
+	 * <@chip_name0{offset|name[=vname]:offset|name[=vname]:...}
+	 * [@chip_name1{offset|name[=vname]:offset|name[=vname]:...}]
+	 * [@chip_name2{offset|name[=vname]:offset|name[=vname]:...}]
+	 * ...>
+	 */
+
+	b = o = strdup(opts);
+	while ((tmp = strsep(&o, "@")) != NULL) {
+
+		/* discard subsequent chips */
+		if (cn >= VIRTIO_GPIO_MAX_CHIPS ||
+				ln >= VIRTIO_GPIO_MAX_VLINES) {
+			DPRINTF("gpio chips reach max:%d\n", cn);
```

Comment:
Could we break out of this loop if `ln >= VIRTIO_GPIO_MAX_VLINES`? If so, can we add this to the error message to make it more complete (i.e. print out both `cn` and `ln` parameters)?

Reply:
Sure,  need to add complete log information, thanks